lib/matplotlib/ticker.py

```diff
@@ -2502,64 +2685,116 @@ def view_limits(self, vmin, vmax):
+    def minor(self, value):
+        self.set_params(minor=value)
 
     def tick_values(self, vmin, vmax):
         # dummy axis has no axes attribute
-        if hasattr(self.axis, 'axes') and self.axis.axes.name == 'polar':
-            raise NotImplementedError('Polar axis cannot be logit scaled yet')
-
-        vmin, vmax = self.nonsingular(vmin, vmax)
-        vmin = np.log10(vmin / (1 - vmin))
-        vmax = np.log10(vmax / (1 - vmax))
+        if hasattr(self.axis, "axes") and self.axis.axes.name == "polar":
+            raise NotImplementedError("Polar axis cannot be logit scaled yet")
 
-        decade_min = np.floor(vmin)
-        decade_max = np.ceil(vmax)
-
-        # major ticks
-        if not self.minor:
-            ticklocs = []
-            if decade_min <= -1:
-                expo = np.arange(decade_min, min(0, decade_max + 1))
-                ticklocs.extend(10**expo)
-            if decade_min <= 0 <= decade_max:
-                ticklocs.append(0.5)
-            if decade_max >= 1:
-                expo = -np.arange(max(1, decade_min), decade_max + 1)
-                ticklocs.extend(1 - 10**expo)
-
-        # minor ticks
+        if self._nbins == "auto":
+            if self.axis is not None:
+                nbins = self.axis.get_tick_space()
+                if nbins < 2:
+                    nbins = 2
+            else:
+                nbins = 9
         else:
-            ticklocs = []
-            if decade_min <= -2:
-                expo = np.arange(decade_min, min(-1, decade_max))
-                newticks = np.outer(np.arange(2, 10), 10**expo).ravel()
-                ticklocs.extend(newticks)
-            if decade_min <= 0 <= decade_max:
-                ticklocs.extend([0.2, 0.3, 0.4, 0.6, 0.7, 0.8])
-            if decade_max >= 2:
-                expo = -np.arange(max(2, decade_min), decade_max + 1)
-                newticks = 1 - np.outer(np.arange(2, 10), 10**expo).ravel()
-                ticklocs.extend(newticks)
+            nbins = self._nbins
 
-        return self.raise_if_exceeds(np.array(ticklocs))
+        # We define ideal ticks with their index:
+        # linscale: ... 1e-3 1e-2 1e-1 1/2 1-1e-1 1-1e-2 1-1e-3 ...
+        # b-scale : ... -3   -2   -1   0   1      2      3      ...
+        def ideal_ticks(x):
+            return 10 ** x if x < 0 else 1 - (10 ** (-x)) if x > 0 else 1 / 2
+
+        vmin, vmax = self.nonsingular(vmin, vmax)
+        binf = int(
+            np.floor(np.log10(vmin))
+            if vmin < 0.5
+            else 0
+            if vmin < 0.9
+            else -np.ceil(np.log10(1 - vmin))
+        )
+        bsup = int(
+            np.ceil(np.log10(vmax))
+            if vmax <= 0.5
+            else 1
+            if vmax <= 0.9
+            else -np.floor(np.log10(1 - vmax))
+        )
+        numideal = bsup - binf - 1
+        if numideal >= 2:
+            # have 2 or more wanted ideal ticks, so use them as major ticks
+            if numideal > nbins:
```

Comment:
I think(?) you can perhaps get rid of this close and take advantage of the fact that if numideal <= nbins, then subsampling_factor = math.ceil(numideal / nbins) == 1 so there's no subsampling.

Also it *may* be easier to follow if you split apart the major and minor cases first?  Right now it's not that easy to follow the major/minor cases.

Just some suggestions, nothing compulsory.

Reply:
Maybe it because I did this code, I understand what you suggest, but the resulting code seem more difficult to understand. But I wrote this code, I am partial.